keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,191 @@
+    - [Reflector](#reflector)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Consistent reads may be served from cache so long as:
+- A consistent (quorum) read is first made to etcd to get the latest "revsion"
+- The data in the watch cache no older than the latest "revsion" just from etcd
+
+etcd 3.4 contains a [progress notify](https://github.com/etcd-io/etcd/issues/9855)
+feature that was added specifically to make it easy to check if a cache that is
+updated via an etcd watch is up-to-date relative to some etcd revsion.
+
+This KEP summarizes how we can take advantage of progress notify feature in etcd
+3.4 to efficiently determine how up-to-date kubernetes watch caches are then
+serve reads from the watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Intuitively, we expect that serving reads from the watch cache to be more
+performant and scalable than reading them from etcd, deserializing them,
+converting them to the desired type and then garbage collecting all the
+objects that were allocated during the read. 
+
+We will need to measure the impact to performance and scalability, but we
+believe there is a significant opportunity and would like to explore it.
+
+In addition to the general scale and performance opportunity, we also aim
+to resolve a specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) is still open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted, which can
+curently happen if the initial list request is served from a stale watch cache
+with data much older than the reflector has previously observed or if the
+api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Serve consistent reads from the watch cache for etcd 3.4 or newer
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scailability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+- Serve consistent reads from the watch cache for etcd 3.3 or earlier
+
+## Proposal
+
+Allow the etcd minor version to be specified in the kube-apiserver `--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  For backward compatibility, if `--storage-backend=etcd3` is provided, it will treat this as `etcd3.0`.
```

Comment:
Would it be possible for the api server to discover the version of etcd on start up?  If the feature goes GA and is enabled by default, it would be nice to not require config changes to use it (i.e. upgrading etcd would be enough).

Reply:
We've briefly discussed this with @jpbetz yesterday. 
How I would see that is:
- introduce a mechanism to control switching it off by a flag (I'm not 100% convinced it should be that flag, but that's a minor detail)
- as you described, periodically detect etcd version and use it to automatically disable the feature if etcd version is too old
- the flag will just be a safety mechanism for the operator, rather than something everyone has to explicitly set

With that, by default operators won't have to care about that at all, but they will have a mechanism to disable just in case.
Assuming the above, I'm slightly concerned with reusing the `--storage-backend` flag to and using it to control multiple different things. But i agree it's not an issue for now, so maybe that's fine.

Reply:
> > I took a cursory glance a question came to my mind. How would this work for paginated list calls? I couldn't find anything in the proposal on that.
> 
> It's not solved. That said, I think that we don't necessary need to address that from day1.
> In my opinion, the biggest gain from it we can get from node-originating requests (e.g. kubelet listing pods scheduled on its node). For those requests, the size of the response is small (it fits a single page, assuming you won't make it extremely small), whereas the number of objects to process is proportional to cluster-size (so fairly big).
> 
> So as an example, in 5k-node cluster with 30pods/node, listing "pods from node X" returns 30 items, but listing from etcd would result in listing 150k pods, deserializing them, filtering and returning those 30. Listing from cache means just filtering and returning those 30, and once [kubernetes/kubernetes#85445](https://github.com/kubernetes/kubernetes/pull/85445) merges (and one more small PR on top of it), it will be proportional to the size of result.

This is a valuable insight. I've added it to the motivation section. (and added to wojtek-t to KEP authors).

> 
> So yes - this doesn't immediately solve all problem, but it may visibly help with many of those.
> 
> I can imagine the algorithm like:
> 
> * if limit is set, check if we set selectors - if not fallback to etcd, if so, assume high probability the result will not be too large
> * ensure cache is up-to-date
> * try to list from cache
> * if result is too large, fallback to list from etcd
> 
> Joe Betz - while the above algorithm is probably not something we want to put in the KEP (do we?), the above explanation I made actually is probably worth putting (parts of it should do into motivation, I think).

Humm, so the problem is that, because the watch cache does not paginate responses, clients that were previously getting pagination for RV="" requests will start getting paginated responses.

I'll comment about this and add it to the list of things we'd need to address before going to beta.

If it turns out to be a real issue, maybe it would best be solved by supporting pagination in the watch cache? I haven't though about the complications around that...